### ISSUE TITLE
Mega Macro Blocked by Blizzard UI

### ISSUE THREAD

**Issue by Magustar** [Sellorio/mega-macro#179]
I've not seen anything in Bug Grabber but it seems like every time I hit a loading screen I have to /reload ui or face the following popup;

![image](https://user-images.githubusercontent.com/73049124/200982122-dd628679-938a-4b69-91df-fab1624161b1.png)

I see errors from EditModeSystemTemplates

![image](https://user-images.githubusercontent.com/73049124/200982360-a34b887a-b914-4421-aae0-fb1f126390d7.png)

It seems like this may be an error in Blizzards new UI but is there any workaround we can do make MegaMacros more stable?

**Comment by Dannez83** [comment]
[mega-macro-action-bar-engine.zip](https://github.com/Sellorio/mega-macro/files/9980233/mega-macro-action-bar-engine.zip)

try replacing the above in MegaMacro\src\engine

Backup the old file first in case there is some other addon you have that conflicts with these changes. I've put in some extra measures for dealing with bartender, but no idea if it has issues with other UI addons. 
Something else to note - this fix is a bit hacky. Icons will break for macros where a toy is used as the icon/tooltip. But small price to pay imo

**Comment by Magustar** [comment]
Dannez83,

Thanks that helped the stability greatly I've not seen any hard crashes but I'm noticing some quirks.  Timers on buttons using #showtooltip [mod:ctrl] or [mod:shift] show the timer of the spell without mod not their own timers.  While I haven't seen any hard crashes I do get bug catcher errors and once I see those my trinkets placed in other action bars are unresponsive without a reload ui

**Comment by Magustar** [comment]
And no sooner did I post this I did get a hard crash, but it does seem to be much less frequent.

**Comment by aurelion314** [comment]
Hey @Magustar, Are you still getting these errors? I haven't had them myself, but I created some options that might fix it. They limit some functionality, but might beat having to reload frequently. Version is [here](https://github.com/aurelion314/mega-macro). I'm trying to get feedback on if the features are worth it before requesting a merge. You have to go to the config tab and enable the options before it does anything different.

**Comment by Magustar** [comment]
Hi @aurelion314 , Yeah I seem to notice some patterns here.  I notice it's much worse when the bars change ie. druids changing form or rogues going in and out of stealth.  If that helps in narrowing down the issue.  I'll check out your link though and let you know.

**Comment by southpaw168** [comment]
Hi guys. I get the blizzard block aswell. I'm playing druid and as Magustar said. It's mainly when i shift or the spell has to change stance. If I can contribute more details let me know

**Comment by Magustar** [comment]
@aurelion314 so tried the branch you linked and didn't see any change.

**Comment by aurelion314** [comment]
@Magustar did you enable the options in the config tab? It does no thing by default, but should start using default blizzard icons/macro once enabled. If it still errors with that, that's quite interesting..

**Comment by JoeLatte88** [comment]
@aurelion314 , I took the code from your fork and I don't seem to be getting actions blocks anymore. Used to happen every couple minutes, but now doesn't occur after 30 minutes of constant combat..... I'll keep testing, but looks good so far.

**Comment by Magustar** [comment]
@aurelion314 Hmm maybe that's what I'm missing?  How do I enable the config tab?  I don't see any options in game.

**Comment by aurelion314** [comment]
It should simply appear as another tab:
![image](https://user-images.githubusercontent.com/23002984/204934407-44d3a18e-658e-455a-b772-dab817f3d9e2.png)

**Comment by Magustar** [comment]
HAH! You're absolutely right.  I was looking in the addon folder and in the files.  Thanks.

**Comment by Dannez83** [comment]
@aurelion314 Do the config settings reintroduce the blizzard limit on the number of macros per character, or just the character limit of the macros?

**Comment by aurelion314** [comment]
> @aurelion314 Do the config settings reintroduce the blizzard limit on the number of macros per character, or just the character limit of the macros?

Just the character limit. You'll still have the spec/class tabs and it doesn't affect the saved Macro data. So if you have a 1000 character macro, it will still be there if you switch back to Mega Macro style. It is simply truncated while using Bliz style.

**Comment by Dannez83** [comment]
> > @aurelion314 Do the config settings reintroduce the blizzard limit on the number of macros per character, or just the character limit of the macros?
> 
> Just the character limit. You'll still have the spec/class tabs and it doesn't affect the saved Macro data. So if you have a 1000 character macro, it will still be there if you switch back to Mega Macro style. It is simply truncated while using Bliz style.
> 
> I now realize that 'Blizzard Style' may be poor word choice, since it might sound as if it will revert entirely to blizzard UI and everything. However it only changes how the macros are bound behind the scenes.

ok cheers. I have a lot of macros on but they are all quite short, so I will give this a go and report back

**Comment by durandal42** [comment]
> I'm trying to get feedback on if the features are worth it before requesting a merge.

For me the *only* critical feature of MegaMacro is macro sharing between characters in excess of the default limit on global macros, so I'll definitely give this a try.

**Comment by southpaw168** [comment]
The Blizzard Action Bar Icons options fixes the UI Block. But all my macros are questionmarks now.

**Comment by YetAnotherDevWannabe** [comment]
Thanks for the the option @aurelion314.
So in the end we still have no clue what is creating the problem and how to solve it ?

**Comment by YetAnotherDevWannabe** [comment]
Even with your edits I keep having the issue, but it's way more rare.
It seems to be linked to grouping, but not sure.

**Comment by gillouche** [comment]
hi @aurelion314, your branch completely fixed all my issue with Mega Macro. Would it be possible to have it merged in the main repo? Thanks!

**Comment by Sydafex** [comment]
As Magustar mentioned, this issue is quite common again- any updates on this?

**Comment by durandal42** [comment]
Apparently I've been running @aurelion314's version for quite some months now, and so the latest update broke me again.

Can we get aurelion's fixes merged into the master branch?

**Comment by aurelion314** [comment]
@durandal42 do you use long macros or rely on Mega Macro's icon engine? I've been debating whether a highly stripped down version of Mega Macro that just does the macro organization might be worthwhile. I have ideas of how to avoid overwriting existing macros too. I just returned this season and could use Mega Macro again so I've been thinking about it.

**Comment by durandal42** [comment]
@aurelion314 The *only* load-bearing feature, for me, is the organization. Specifically, I really like being able to share macros across different characters with the same class, without running into the limit on account-level macros.

**Comment by aurelion314** [comment]
Ok same thing for me. If I can find the time I'll see about isolating just that part.

**Comment by aurelion314** [comment]
@durandal42 
If you want to try it out, I'm working on some changes here: https://github.com/aurelion314/mega-macro/tree/Engine-Rework

- It will now default to using native blizzard bindings whenever the macro is below 250 characters, thus avoiding potential faulty /click behaviour. 
- Improved imports. It will no longer fail imports, but will move excess macros to a new 'Inactive' tab where you can then copy or delete them. 
- Fixing any bugs I find, which hasn't actually been too many. Seems surprisingly stable :)
- Merged character and character specialization. I needed the tab and I never found character spec to be useful. 

Note that it isn't polished just yet. Could be some bugs, and since I did rework some import logic, might be best to back up WTF folder first... Unlikely to be necessary, but good to be safe ;)
